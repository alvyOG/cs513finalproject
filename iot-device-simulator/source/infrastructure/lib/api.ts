
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, Construct, RemovalPolicy } from '@aws-cdk/core';
import {
  AccessLogFormat,
  AuthorizationType,
  ContentHandling,
  Deployment,
  EndpointType,
  Integration,
  IntegrationType,
  LogGroupLogDestination,
  MethodLoggingLevel,
  MethodOptions,
  PassthroughBehavior,
  RequestValidator,
  RestApi,
  Stage
} from '@aws-cdk/aws-apigateway';
import { Policy, ServicePrincipal } from '@aws-cdk/aws-iam';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { addCfnSuppressRules } from '../utils/utils';
import { Function as LambdaFunction } from '@aws-cdk/aws-lambda';
import { IBucket } from '@aws-cdk/aws-s3';
import { Table } from '@aws-cdk/aws-dynamodb';

/**
 * ApiConstructProps props
 * @interface ApiConstructProps
 */
export interface ApiConstructProps {
  // Policy for CloudWatch Logs
  readonly cloudWatchLogsPolicy: Policy;
  // IoT endpoint address
  readonly stepFunctionsARN: string;
  // Simulation data DynamoDB table name
  readonly simulationTable: Table;
  // Device Type data DynamoDB table name
  readonly deviceTypeTable: Table;
  //Routes S3 bucket
  readonly routesBucketArn: string;
  //microservices lambda
  readonly microservicesLambda: LambdaFunction;
  /**
   * Solution config properties.
   * Logging level, solution ID, version, source code bucket, and source code prefix
   */
  readonly solutionConfig: {
    sendAnonymousUsage: string;
    solutionId: string;
    solutionVersion: string;
    sourceCodeBucket: IBucket;
    sourceCodePrefix: string;
  };
  // Solution UUID
  readonly uuid: string;
}

/**
 * @class
 * IoT Device Simulator Framework API Construct.
 * It creates an API Gateway REST API and other resources.
 */
export class ApiConstruct extends Construct {
  // API endpoint
  public apiEndpoint: string;
  // API ID
  public apiId: string;
  //microservices lambda
  public microservicesLambdaFunction: LambdaFunction;

  constructor(scope: Construct, id: string, props: ApiConstructProps) {
    super(scope, id);

    const apiLogGroup = new LogGroup(this, 'Logs', {
      removalPolicy: RemovalPolicy.DESTROY,
      retention: RetentionDays.THREE_MONTHS
    });
    addCfnSuppressRules(apiLogGroup, [{
      id: 'W84', reason: 'CloudWatch Logs are already encrypted by default.'
    }]);

    const api = new RestApi(this, 'IoTDeviceSimulatorApi', {
      defaultCorsPreflightOptions: {
        allowOrigins: ['*'],
        allowHeaders: [
          'Authorization',
          'Content-Type',
          'X-Amz-Date',
          'X-Amz-Security-Token',
          'X-Api-Key'
        ],
        allowMethods: [
          'GET',
          'POST',
          'PUT',
          'DELETE',
          'OPTIONS'
        ],
        statusCode: 200
      },
      deploy: true,
      deployOptions: {
        accessLogDestination: new LogGroupLogDestination(apiLogGroup),
        accessLogFormat: AccessLogFormat.jsonWithStandardFields(),
        loggingLevel: MethodLoggingLevel.INFO,
        stageName: 'prod',
        tracingEnabled: true
      },
      description: 'IoT Device Simulator Rest API',
      endpointTypes: [EndpointType.REGIONAL]
    });
    this.apiEndpoint = `https://${api.restApiId}.execute-api.${Aws.REGION}.amazonaws.com/prod`;
    this.apiId = api.restApiId;

    const requestValidator = new RequestValidator(this, 'ApiRequestValidator', {
      restApi: api,
      validateRequestParameters: true,
      validateRequestBody: true
    });

    addCfnSuppressRules(api.node.findChild('Deployment') as Deployment, [{
      id: 'W68', reason: 'The solution does not require the usage plan.'
    }]);
    addCfnSuppressRules(api.node.findChild('DeploymentStage.prod') as Stage, [{
      id: 'W64', reason: 'The solution does not require the usage plan.'
    }]);

    /**
     * method options for all methods
     */
    const universalMethodOptions: MethodOptions = {
      authorizationType: AuthorizationType.IAM,
      methodResponses: [{
        statusCode: '200',
        responseModels: {
          'application/json': { modelId: 'Empty' }
        }
      }],
      requestParameters: { 'method.request.querystring.nextToken': false },
      requestValidator: requestValidator
    };

    /**
     * Integration for all resources
     */
    const universalIntegration = new Integration({
      type: IntegrationType.AWS_PROXY,
      integrationHttpMethod: 'POST',
      options: {
        contentHandling: ContentHandling.CONVERT_TO_TEXT,
        integrationResponses: [{ statusCode: '200' }],
        passthroughBehavior: PassthroughBehavior.WHEN_NO_MATCH
      },
      uri: `arn:${Aws.PARTITION}:apigateway:${Aws.REGION}:lambda:path/2015-03-31/functions/${props.microservicesLambda.functionArn}/invocations`
    });

    /**
     * simulation API
     * ANY /simulation
     * ANY /simulation/{simid}
     */
    const simulationResource = api.root.addResource('simulation');
    simulationResource.addMethod(
      'ANY',
      universalIntegration,
      universalMethodOptions
    );
    const simulationSimIdResource = simulationResource.addResource('{simid}');
    simulationSimIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    /**
     * Devive Types API
     * ANY /devicetypes
     * ANY /devicetypes/{typeid}
     */
    const deviceTypesResource = api.root.addResource('devicetypes');
    deviceTypesResource.addMethod('ANY', universalIntegration, universalMethodOptions);
    const typeIdResource = deviceTypesResource.addResource('{typeid}');
    typeIdResource.addMethod('ANY', universalIntegration, universalMethodOptions);

    props.microservicesLambda.addPermission('ApiLambdaInvokePermission', {
      action: 'lambda:InvokeFunction',
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      sourceArn: api.arnForExecuteApi()
    });
  }
}