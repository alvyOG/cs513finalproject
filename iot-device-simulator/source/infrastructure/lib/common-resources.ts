// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { ArnFormat, Construct, RemovalPolicy, Stack } from '@aws-cdk/core';
import { Effect, Policy, PolicyStatement } from '@aws-cdk/aws-iam';
import { BlockPublicAccess, Bucket, BucketAccessControl, BucketEncryption, IBucket } from '@aws-cdk/aws-s3';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * CommonResourcesConstruct props
 * @interface CommonResourcesConstructProps
 */
export interface CommonResourcesConstructProps {
  // Source code bucket
  sourceCodeBucket: string;
}

/**
 * @class
 * IoT Device Simulator Common Resources Construct.
 * It creates a common CloudWatch Logs policy for Lambda functions, a logging S3 bucket, and M2C2 S3 bucket.
 */
export class CommonResourcesConstruct extends Construct {
  // CloudWatch Logs policy
  public cloudWatchLogsPolicy: Policy;
  // S3 Logging bucket
  public s3LoggingBucket: Bucket;
  // Code S3 bucket
  public sourceCodeBucket: IBucket;

  constructor(scope: Construct, id: string, props: CommonResourcesConstructProps) {
    super(scope, id);

    this.cloudWatchLogsPolicy = new Policy(this, 'CloudWatchLogsPolicy', {
      statements: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'logs:CreateLogGroup',
            'logs:CreateLogStream',
            'logs:PutLogEvents'
          ],
          resources: [
            Stack.of(this).formatArn({ service: 'logs', resource: 'log-group', resourceName: '/aws/lambda/*', arnFormat: ArnFormat.COLON_RESOURCE_NAME })
          ]
        })
      ]
    });

    this.s3LoggingBucket = new Bucket(this, 'LogBucket', {
      accessControl: BucketAccessControl.LOG_DELIVERY_WRITE,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
      encryption: BucketEncryption.S3_MANAGED,
      removalPolicy: RemovalPolicy.RETAIN
    });

    addCfnSuppressRules(this.s3LoggingBucket, [
      { id: 'W35', reason: 'This bucket is to store S3 logs, so it does not require access logs.' },
      { id: 'W51', reason: 'This bucket is to store S3 logs, so it does not require S3 policy.' }
    ]);

    this.sourceCodeBucket = Bucket.fromBucketName(this, 'SourceCodeBucket', props.sourceCodeBucket);
  }
}