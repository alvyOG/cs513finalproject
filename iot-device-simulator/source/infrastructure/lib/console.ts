// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Aws, ArnFormat, Construct, Duration, RemovalPolicy, Stack } from '@aws-cdk/core';
import { CfnIdentityPool, CfnIdentityPoolRoleAttachment, CfnUserPool, CfnUserPoolUser, UserPool, UserPoolClient } from '@aws-cdk/aws-cognito';
import { Effect, FederatedPrincipal, PolicyDocument, PolicyStatement, Role } from '@aws-cdk/aws-iam';
import { Bucket, IBucket } from '@aws-cdk/aws-s3';
import { CloudFrontToS3 } from '@aws-solutions-constructs/aws-cloudfront-s3';
import { CfnPolicy } from '@aws-cdk/aws-iot';
import { addCfnSuppressRules } from '../utils/utils';

/**
 * UiConstructProps props
 * @interface ConsoleConstructProps
 */
export interface ConsoleConstructProps {
  // The API ID
  readonly apiId: string;
  // S3 logging bucket
  readonly s3LogsBucket: Bucket;
  // User E-Mail address
  readonly adminEmail: string;
  //map resources
  readonly mapArn: string;
  readonly placeIndexArn: string;
}

/**
 * @class
 * IoT Device Simulator Framework UI Construct.
 * It creates a CloudFront distribution, an UI hosting S3 bucket,
 * Cognito resources, and custom resources for UI assets.
 */
export class ConsoleConstruct extends Construct {
  // CloudFront distribution domain name
  public cloudFrontDomainName: string;
  // Cognito Identity pool ID
  public identityPoolId: string;
  // UI S3 bucket
  public consoleBucket: IBucket;
  // Cognito user pool ID
  public userPoolId: string;
  // Cognito user pool web client ID
  public webClientId: string;
  //IoT policy name
  public iotPolicy: CfnPolicy;

  constructor(scope: Construct, id: string, props: ConsoleConstructProps) {
    super(scope, id);

    const consoleCloudfrontDist = new CloudFrontToS3(this, 'Distribution', {
      bucketProps: {
        serverAccessLogsBucket: props.s3LogsBucket,
        serverAccessLogsPrefix: 'console-s3/'
      },
      cloudFrontDistributionProps: {
        comment: 'IoT Device Simulator Distribution',
        enableLogging: true,
        errorResponses: [
          { httpStatus: 403, responseHttpStatus: 200, responsePagePath: '/index.html' },
          { httpStatus: 404, responseHttpStatus: 200, responsePagePath: '/index.html' }
        ],
        httpVersion: 'http2',
        logBucket: props.s3LogsBucket,
        logFilePrefix: 'console-cf/'

      },
      insertHttpSecurityHeaders: false
    });
    this.cloudFrontDomainName = consoleCloudfrontDist.cloudFrontWebDistribution.domainName;
    this.consoleBucket = consoleCloudfrontDist.s3BucketInterface;

    const userPool = new UserPool(this, 'UserPool', {
      passwordPolicy: {
        minLength: 12,
        requireDigits: true,
        requireLowercase: true,
        requireSymbols: true,
        requireUppercase: true
      },
      removalPolicy: RemovalPolicy.DESTROY,
      selfSignUpEnabled: false,
      signInAliases: {
        email: true
      },
      userPoolName: `${Aws.STACK_NAME}-user-pool`,
      userInvitation: {
        emailSubject: '[IoT Device Simulator] Login information',
        emailBody: `
          <p>
            You are invited to join IoT Device Simulator.<br />
            https://${this.cloudFrontDomainName}
          </p>
          <p>
            Please sign in to IoT Device Simulator using the temporary credentials below:<br />
            Username: <strong>{username}</strong><br />Password: <strong>{####}</strong>
          </p>
        `
      }
    });
    (userPool.node.defaultChild as CfnUserPool).userPoolAddOns = { advancedSecurityMode: 'ENFORCED' };
    this.userPoolId = userPool.userPoolId;

    const userPoolClient = new UserPoolClient(this, 'UserPoolClient', {
      generateSecret: false,
      preventUserExistenceErrors: true,
      refreshTokenValidity: Duration.days(1),
      userPool,
      userPoolClientName: `${Aws.STACK_NAME}-userpool-client`
    });
    this.webClientId = userPoolClient.userPoolClientId;

    new CfnUserPoolUser(this, 'User', { // NOSONAR: typescript:S1848
      userPoolId: userPool.userPoolId,
      desiredDeliveryMediums: ['EMAIL'],
      forceAliasCreation: true,
      userAttributes: [
        { name: 'email', value: props.adminEmail },
        { name: 'email_verified', value: 'true' }
      ],
      username: props.adminEmail
    });

    const identityPool = new CfnIdentityPool(this, 'IdentityPool', {
      allowUnauthenticatedIdentities: false,
      cognitoIdentityProviders: [{
        clientId: userPoolClient.userPoolClientId,
        providerName: userPool.userPoolProviderName,
        serverSideTokenCheck: false
      }]
    });


    this.identityPoolId = identityPool.ref;

    const idsIotPolicy = new CfnPolicy(this, 'IDS-IoT-Policy', {
      policyDocument: new PolicyDocument({
        statements: [
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Connect"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'client', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Subscribe"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topicfilter', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          }),
          new PolicyStatement({
            effect: Effect.ALLOW,
            actions: ["iot:Receive"],
            resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topic', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
          })
        ]
      })
    });

    this.iotPolicy = idsIotPolicy;
    addCfnSuppressRules(idsIotPolicy, [
      { id: 'W11', reason: 'Cannot specify the resource to attach policy to identity.' }
    ]);

    const authenticatedRole = new Role(this, 'IdentityPoolAuthenticatedRole', {
      assumedBy: new FederatedPrincipal(
        'cognito-identity.amazonaws.com',
        {
          StringEquals: { 'cognito-identity.amazonaws.com:aud': identityPool.ref },
          'ForAnyValue:StringLike': { 'cognito-identity.amazonaws.com:amr': 'authenticated' }
        },
        'sts:AssumeRoleWithWebIdentity'
      ),
      description: `${Aws.STACK_NAME} Identity Pool authenticated role`,
      inlinePolicies: {
        'ExecuteApiPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ['execute-api:Invoke'],
              resources: [Stack.of(this).formatArn({ service: 'execute-api', resource: `${props.apiId}`, resourceName: 'prod/*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
          ],
        }),
        'LocationServicePolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "geo:SearchPlaceIndexForText",
                "geo:GetMapGlyphs",
                "geo:GetMapSprites",
                "geo:GetMapStyleDescriptor",
                "geo:SearchPlaceIndexForPosition",
                "execute-api:Invoke",
                "geo:GetMapTile"
              ],
              resources: [
                props.mapArn,
                props.placeIndexArn,
              ]
            }),
          ]
        }),
        'IoTPolicy': new PolicyDocument({
          statements: [
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: [
                "iot:AttachPrincipalPolicy"
              ],
              resources: ['*']
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Connect"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'client', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Subscribe"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topicfilter', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            }),
            new PolicyStatement({
              effect: Effect.ALLOW,
              actions: ["iot:Receive"],
              resources: [Stack.of(this).formatArn({ service: 'iot', resource: 'topic', resourceName: '*', arnFormat: ArnFormat.SLASH_RESOURCE_NAME })]
            })
          ]
        })
      }
    });
    addCfnSuppressRules(authenticatedRole, [
      { id: 'W11', reason: 'iot:AttachPrincipalPolicy does not allow for resource specification' }
    ]);

    new CfnIdentityPoolRoleAttachment(this, 'IdentityPoolRoleAttachement', { // NOSONAR: typescript:S1848
      identityPoolId: identityPool.ref,
      roles: { authenticated: authenticatedRole.roleArn }
    });
  }
}