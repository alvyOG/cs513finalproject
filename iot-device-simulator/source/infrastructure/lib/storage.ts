// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import { Construct, RemovalPolicy } from '@aws-cdk/core';
import { AttributeType, BillingMode, Table, TableEncryption } from '@aws-cdk/aws-dynamodb';
import { Bucket, BucketEncryption, BlockPublicAccess} from '@aws-cdk/aws-s3';
import { PolicyStatement, AnyPrincipal, Effect } from '@aws-cdk/aws-iam';

/**,
 * @interface StorageContructProps
 * StorageContruct props
 */
export interface StorageContructProps {
  readonly solutionId: string;
  readonly s3LogsBucket: Bucket
}

/**
 * IoT Device Simulator storage construct
 * Creates an S3 bucket to store test scenarios and 
 * a DynamoDB table to store tests and test configuration
 */
export class StorageContruct extends Construct {
  public simulationsTable: Table;
  public deviceTypesTable: Table;
  public routesBucket: Bucket

  constructor(scope: Construct, id: string, props: StorageContructProps) {
    super(scope, id);

    this.simulationsTable = new Table(this, 'IDS-Simulations-Table', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: 'simId', type: AttributeType.STRING },
      pointInTimeRecovery: true
    });

    this.deviceTypesTable = new Table(this, 'IDS-Device-Types-Table', {
      billingMode: BillingMode.PAY_PER_REQUEST,
      encryption: TableEncryption.AWS_MANAGED,
      partitionKey: { name: 'typeId', type: AttributeType.STRING },
      pointInTimeRecovery: true
    });

    this.routesBucket = new Bucket(this, 'RoutesBucket', {
      removalPolicy: RemovalPolicy.RETAIN,
      serverAccessLogsBucket: props.s3LogsBucket,
      serverAccessLogsPrefix: 'routes-bucket-access/',
      encryption: BucketEncryption.S3_MANAGED,
      blockPublicAccess: BlockPublicAccess.BLOCK_ALL,
    });

    this.routesBucket.addToResourcePolicy(
      new PolicyStatement({
        sid: 'HttpsOnly',
        resources: [
          `${this.routesBucket.bucketArn}/*`,
          `${this.routesBucket.bucketArn}`
        ],
        actions: ['*'],
        principals: [new AnyPrincipal()],
        effect: Effect.DENY,
        conditions:
        {
          Bool: {
            'aws:SecureTransport': 'false'
          }
        }
      })
    );
  }
}