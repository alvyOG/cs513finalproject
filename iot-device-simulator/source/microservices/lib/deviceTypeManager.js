// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * @author Solution Builders
 */

'use strict';
const { sendAnonymousMetric } = require('../metrics/index');
const AWS = require('aws-sdk');
const { nanoid } = require('nanoid');
const { SOLUTION_ID, VERSION } = process.env;
let options = {};
if (SOLUTION_ID && VERSION && SOLUTION_ID.trim() && VERSION.trim()) {
    options.customUserAgent = `AwsSolution/${SOLUTION_ID}/${VERSION}`;
}
let docClient = new AWS.DynamoDB.DocumentClient(options);

/**
 * Performs crud actions for a device type, such as, creating, retrieving, updating and deleting device types.
 *
 * @class DeviceTypeManager
 */
class DeviceTypeManager {
    /**
     * Get device types for the user.
     *
     */
    async getDeviceTypes() {
        const params = {
            TableName: process.env.DEVICE_TYPES_TBL,
        };
        try {
            let results = await docClient.scan(params).promise();
            let lastEvalKey = results.LastEvaluatedKey;
            while (lastEvalKey && Object.keys(lastEvalKey).length > 0) {
                params.ExclusiveStartKey = lastEvalKey;
                let newResults = await docClient.scan(params).promise();
                results.Items.push(...newResults.Items);
                lastEvalKey = newResults.LastEvaluatedKey;
            }
            return (results.Items);
        } catch (err) {
            console.error(err);
            console.error(`Error occurred while attempting to retrieve device types.`);
            throw (err);
        }
    }

    /**
     * Creates a device type for user.
     * @param {object} deviceType - device type object
     */
    async createDeviceType(deviceType) {
        try {
            let _id;
            if (deviceType.typeId && deviceType.typeId !== 'idsAutoDemo') {
                _id = deviceType.typeId;
            } else {
                const suffix = deviceType.typeId || ""
                _id = nanoid(9) + suffix;
            }
            const date = new Date().toISOString();
            let _deviceType = {
                typeId: _id,
                name: deviceType.name,
                topic: deviceType.topic,
                payload: deviceType.payload,
                createdAt: date,
                updatedAt: date
            };
            let params = {
                TableName: process.env.DEVICE_TYPES_TBL,
                Item: _deviceType
            };
            await docClient.put(params).promise();
            if (process.env.SEND_ANONYMOUS_METRIC === 'Yes') {
                let metricData = {
                    eventType: 'create device type',
                };
                metricData.uniquePayloadAttrs = deviceType.payload.reduce((acc, curValue) => {
                    if (!acc.includes(curValue.type)) acc.push(curValue.type);
                    return acc;
                }, []);
                await sendAnonymousMetric(metricData);
            }
            return (_deviceType);
        } catch (err) {
            console.error(err);
            console.error(`Error occurred while attempting to create device type.`);
            throw (err);
        }
    }


    /**
     * Deletes a device type for user.
     * @param {string} deviceTypeId - id of device type to delete
     */
    async deleteDeviceType(deviceTypeId) {
        try {
            let params = {
                TableName: process.env.DEVICE_TYPES_TBL,
                Key: {
                    typeId: deviceTypeId
                }
            };
            let deviceType = await docClient.get(params).promise();
            if (deviceType.Item) {
                return docClient.delete(params).promise();
            } else {
                let error = new Error();
                error.code = 400;
                error.error = 'MissingDeviceType';
                error.message = `The requested device type ${deviceTypeId} does not exist.`;
                throw (error);
            }
        } catch (err) {
            console.error(err);
            console.error(`Error occurred while attempting to delete device type.`);
            throw (err);
        }
    }
}

module.exports = DeviceTypeManager;