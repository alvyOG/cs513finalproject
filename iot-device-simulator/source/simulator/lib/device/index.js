// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * @author Solution Builders
 */
'use strict';
const AWS = require('aws-sdk');
const moment = require('moment');
const Random = require('./generators/random/generator');
const Vehicle = require('./generators/vehicle/generator.js');
let awsOptions = { endpoint: process.env.IOT_ENDPOINT };
const { SOLUTION_ID, VERSION } = process.env;
if (SOLUTION_ID && VERSION && SOLUTION_ID.trim() && VERSION.trim()) {
    awsOptions.customUserAgent = `AwsSolution/${SOLUTION_ID}/${VERSION}`;
}
const iotData = new AWS.IotData(awsOptions);
/**
 * @Class Device - represents a single device
 */
class Device {

    constructor(options, sim, device) {
        this.options = options;
        this.id = device.id;
        this.started = device.started || moment().toISOString();
        this.simId = sim.simId;
        this.sendInterval = null;
        this.stage = device.stage || sim.stage;
        if (device.generator) {
            this.options.currentState = device.generator.currentState;
            this.options.staticValues = device.generator.staticValues;
        }
        this.generator = sim.simId.includes('idsAutoDemo') ? new Vehicle(this.options) : new Random(this.options);
        this.payload = device.payload;
        this.topic = device.topic;
        this.duration = sim.duration * 1000;
        this.interval = sim.interval * 1000;
    }

    /**
     * generates and sends a message on the given interval
     */
    sendOnInterval = () => {
        return (new Promise((resolve, reject) => {
            //generate message on interval
            this.sendInterval = setInterval(async () => {
                //generate the message and publish
                this._generateMessage();
                if (this.stage === "sleeping") {
                    //stop device if no longer running
                    this.generator.stop();
                    clearInterval(this.sendInterval);
                    resolve("complete")
                } else if (this.options.context.getRemainingTimeInMillis() <= this.interval) {
                    //stop device if not enough time to run another interval
                    this.generator.stop();
                    clearInterval(this.sendInterval);
                    //return information necessary to restart device
                    resolve({
                        stage: this.stage,
                        started: this.started,
                        id: this.id,
                        generator: {
                            currentState: this.generator.currentState,
                            staticValues: this.generator.staticValues
                        }
                    });
                }
            }, this.interval);
        }));
    }

    /**
     * Starts the device
     * @returns the result of the device run
     */
    async run() {
        this.stage = "running";
        try {
            //start generating messages on given interval
            return this.sendOnInterval();
        } catch (err) {
            throw (err)
        }
    }

    /**
     * Stops the device
     */
    stop() {
        this.stage = 'sleeping';
    }

    /**
     * publishes the message payload to the given IoT topic
     * @param {string} topic 
     * @param {object} messagePayload 
     * @returns the data from the publish call
     */
    async _publishMessage(topic, messagePayload) {
        //set iot publish params
        let params = {
            topic: topic,
            payload: messagePayload,
            qos: 0
        };
        //publish to IoT topic
        try {
            return iotData.publish(params).promise();
        } catch (err) {
            throw (err);
        }
    }

    /**
     * Generates the message to be published 
     * and publishes to the given IoT topic
     */
    async _generateMessage() {
        //check if the duration has elapsed
        let timeDelta = moment().diff(moment(this.started));
        if (timeDelta > this.duration || !this.generator.isRunning) {
            console.log(`Device '${this.id}' run duration has elapsed.`);
            this.stop();
        } else {
            //generate the message payload
            this.generator.generateMessagePayload(this.payload, this.topic, this.id);
            //publish all messages
            let messagePromises = this.generator.messages.map((_message) => {
                let _payload = JSON.stringify(_message.payload);
                return (this._publishMessage(_message.topic, _payload));
            })
            //wait for messages to publish then clear the generator messages
            try {
                await Promise.all(messagePromises);
                this.generator.clearMessages();
            } catch (err) {
                throw (err);
            }
        }
    }
}

module.exports = Device;