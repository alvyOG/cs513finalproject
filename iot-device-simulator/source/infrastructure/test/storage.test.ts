// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

import '@aws-cdk/assert/jest';
import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { Bucket } from '@aws-cdk/aws-s3';;
import { StorageContruct } from '../lib/storage';


test('IoT Device Simulator storageConstruct Test', () => {
    const stack = new Stack();
    const testBucket = new Bucket(stack, "testBucket", {});

    const storage = new StorageContruct(stack, 'storage', {
        solutionId: 'testId',
        s3LogsBucket: testBucket
    })

    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});