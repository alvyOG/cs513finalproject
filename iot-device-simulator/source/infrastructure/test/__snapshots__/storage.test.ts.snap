// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IoT Device Simulator storageConstruct Test 1`] = `
Object {
  "Resources": Object {
    "storageIDSDeviceTypesTable01F2060C": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "typeId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "typeId",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "storageIDSSimulationsTable052FC452": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "simId",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "simId",
            "KeyType": "HASH",
          },
        ],
        "PointInTimeRecoverySpecification": Object {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": Object {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "storageRoutesBucketBB9EF62B": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "LoggingConfiguration": Object {
          "DestinationBucketName": Object {
            "Ref": "testBucketDF4D7D1A",
          },
          "LogFilePrefix": "routes-bucket-access/",
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "storageRoutesBucketPolicy68A7F4A8": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "storageRoutesBucketBB9EF62B",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "*",
              "Condition": Object {
                "Bool": Object {
                  "aws:SecureTransport": "false",
                },
              },
              "Effect": "Deny",
              "Principal": Object {
                "AWS": "*",
              },
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::GetAtt": Array [
                          "storageRoutesBucketBB9EF62B",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
                Object {
                  "Fn::GetAtt": Array [
                    "storageRoutesBucketBB9EF62B",
                    "Arn",
                  ],
                },
              ],
              "Sid": "HttpsOnly",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "testBucketDF4D7D1A": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "AccessControl": "LogDeliveryWrite",
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
}
`;
